src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java

```diff
@@ -135,6 +139,52 @@ public Message getReferencedMessage()
         return referencedMessage;
     }
 
+    @Override
+    public long getReferencedMessageIdLong()
+    {
+        if (messageReference == null)
+        {
+            throw new IllegalStateException("Cannot get an id for a non-existent reference.");
+        }
+        return messageReference.getMessageIdLong();
+    }
+
+
+    @Nonnull
+    @Override
+    public String getReferencedMessageId()
+    {
+        if (messageReference == null)
+        {
+            throw new IllegalStateException("Cannot get an id for a non-existent reference.");
+        }
+
+        return Long.toUnsignedString(getReferencedMessageIdLong());
+    }
+
+    @Override
+    public boolean hasReferencedMessage()
+    {
+        return messageReference != null;
+    }
+
+    @Override
+    public RestAction<Message> retrieveReferencedMessage()
+    {
+        if (messageReference == null) {
+            throw new IllegalStateException("Cannot get a retrieve a message for a non-existent reference.");
+        }
+
+        if (referencedMessage != null) {
+            return new CompletedRestAction<>(getJDA(), referencedMessage);
+        }
```

Comment:
I'm not sure about the implications of this.
Retrieve typically means "go get it from the API", thus it is always up-to-date.

The functionality of this method indicates that if we initially constructed the `Message` with a full referenced message then this method cannot be used to get an updated version of that message. The developer would have to know that in some cases this method can be used to get updated versions of the message and other times it cannot be.